Microsoft.Azure.Cosmos/src/Routing/GlobalPartitionFailoverEndpointManagerCore.cs

```diff
@@ -0,0 +1,176 @@
+            () => new ConcurrentDictionary<PartitionKeyRange, PartitionKeyRangeFailoverInfo>());
+
+        public GlobalPartitionFailoverEndpointManagerCore(
+            IGlobalEndpointManager globalEndpointManager)
+        {
+            this.globalEndpointManager = globalEndpointManager ?? throw new ArgumentNullException(nameof(globalEndpointManager));
+        }
+
+        private bool CanUsePartitionLevelFailoverLocations(DocumentServiceRequest request)
+        {
+            if (this.globalEndpointManager.ReadEndpoints.Count <= 1)
+            {
+                return false;
+            }
+
+            if (request.ResourceType == ResourceType.Document || 
+                (request.ResourceType == ResourceType.StoredProcedure && request.OperationType == Documents.OperationType.ExecuteJavaScript))
+            {
+                // Disable for multimaster because it currently 
+                // depends on 403.3 to signal the primary region is backup
+                // and to fail back over
+                if (!this.globalEndpointManager.CanUseMultipleWriteLocations(request))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public override bool TryAddPartitionLevelLocationOverride(
+            DocumentServiceRequest request)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (!this.CanUsePartitionLevelFailoverLocations(request))
+            {
+                return false;
+            }
+
+            if (this.PartitionKeyRangeToLocation.IsValueCreated
+                && this.PartitionKeyRangeToLocation.Value.TryGetValue(
+                    request.RequestContext.ResolvedPartitionKeyRange,
+                    out PartitionKeyRangeFailoverInfo partitionKeyRangeFailover))
+            {
+                DefaultTrace.TraceVerbose("Partition level override. URI: {0}, PartitionKeyRange: {1}",
+                    partitionKeyRangeFailover.Current,
+                    request.RequestContext.ResolvedPartitionKeyRange.Id);
+
+                request.RequestContext.RouteToLocation(partitionKeyRangeFailover.Current);
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Marks the current location unavailable for write
+        /// </summary>
+        public override bool TryMarkEndpointUnavailableForPartitionKeyRange(
+            DocumentServiceRequest request,
+            Uri failedLocation)
+        {
+            if (request == null || failedLocation == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (!this.CanUsePartitionLevelFailoverLocations(request))
+            {
+                return false;
+            }
+
+            if (request?.RequestContext?.ResolvedPartitionKeyRange == null)
```

Comment:
nit: we already validated the request is not null, some maybe
```suggestion
            if (request.RequestContext?.ResolvedPartitionKeyRange == null)
```